lib/flop.ex

```diff
@@ -432,6 +442,169 @@ defmodule Flop do
+    :erlang.binary_to_term(bin)
+  end
+
+  @spec first(Queryable.t(), Repo.t(), Flop.t()) :: {[Schema.t()], Meta.t()}
+  def first(_q, _r, %Flop{limit: limit, order_by: order_by})
+      when is_nil(limit) or is_nil(order_by) do
+    raise ArgumentError,
+      message: "Cursor-based pagination requires limit and order by parameters"
+  end
+
+  def first(q, r, %Flop{limit: limit, offset: nil, order_by: order_by} = flop) do
+    results =
+      q
+      |> query(%{flop | limit: limit + 1})
+      |> r.all()
+
+    next_cursor =
+      if length(results) == limit + 1 do
+        results |> List.last() |> Map.take(order_by) |> encode_cursor()
+      end
+
+    {List.delete_at(results, limit),
+     %Meta{
+       current_offset: nil,
+       current_page: nil,
+       flop: flop,
+       has_next_page?: nil,
+       has_previous_page?: nil,
+       next_cursor: next_cursor,
+       next_offset: nil,
+       next_page: nil,
+       page_size: nil,
+       previous_cursor: nil,
+       previous_offset: nil,
+       previous_page: nil,
+       total_count: nil,
+       total_pages: nil
+     }}
+  end
+
+  @spec apply_cursor(Queryable.t(), map(), [order_direction()]) :: Queryable.t()
+  defp apply_cursor(q, cursor, ordering) do
+    Enum.reduce(ordering, q, fn {direction, field}, q ->
+      cond do
+        Enum.member?([:asc, :asc_nulls_first, :asc_nulls_last], direction) ->
+          Query.where(q, [r], field(r, ^field) >= ^cursor[field])
+
+        Enum.member?([:desc, :desc_nulls_first, :desc_nulls_last], direction) ->
+          Query.where(q, [r], field(r, ^field) <= ^cursor[field])
+
+        true ->
+          raise ArgumentError, message: "Invalid order direction supplied"
+      end
+    end)
+  end
+
+  @spec next(Queryable.t(), Repo.t(), Meta.t()) :: {[Schema.t()], Meta.t()}
+  def next(_, _, %Meta{next_cursor: nil} = meta), do: {[], meta}
+
+  def next(
+        q,
+        r,
+        %Meta{
+          flop:
+            %Flop{
+              limit: limit,
+              offset: nil,
+              order_by: order_by,
+              order_directions: order_directions
+            } = flop,
+          next_cursor: next_cursor
+        } = meta
+      ) do
+    cursor = decode_cursor(next_cursor)
+
+    q = apply_cursor(q, cursor, prepare_order(order_by, order_directions))
+
+    results =
+      q
+      |> query(%{flop | limit: limit + 1})
+      |> r.all()
+
+    new_next_cursor =
+      if length(results) == limit + 1 do
+        results |> List.last() |> Map.take(order_by) |> encode_cursor()
+      end
+
+    {List.delete_at(results, limit),
+     %{meta | previous_cursor: next_cursor, next_cursor: new_next_cursor}}
+  end
+
+  @spec reverse_ordering([order_direction()]) :: [order_direction()]
+  defp reverse_ordering(order_directions) do
+    Enum.map(order_directions, fn {order_direction, field} ->
+      {case order_direction do
+         :asc -> :desc
+         :asc_nulls_first -> :desc_nulls_first
```

Comment:
I think `:asc_nulls_first` would turn to `:desc_nulls_last` etc., no?

Reply:
Ah yeah, I think you're right. I'll be sure to add a test for that.

Reply:
Technically, it doesn't even make sense to have null values in a cursor column anyway. But that's another point: We also need to think about how to handle ordering with cursor-based pagination. Probably there needs to be a way to specify which columns may be used for the cursor, and which one to use in a query. Passing any other order params should be invalid.